File: leetcode/Python/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def traverse(cur):
            if cur is None:
                return 0
            l = traverse(cur.left)
            r = traverse(cur.right)
            sol[0] = max(sol[0], l+r)
            return max(l, r) + 1
        sol = [0]
        traverse(root)
        return sol[0]

File: leetcode/C++/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int traverse(TreeNode* cur, int &sol){
        if(cur==nullptr) return 0;
        int l=traverse(cur->left, sol);
        int r=traverse(cur->right, sol);
        sol=max(sol, l+r);
        return max(l, r)+1;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int sol=0;
        traverse(root, sol);
        return sol;
    }
};